### ISSUE TITLE
Radio-buttons out of sync with model (reactive-forms)

### ISSUE THREAD

**Issue by Warwizza** [angular/angular#38196]
🐞 bug report

### Is this a regression?
Not that I know

### Description
I'm creating a survey app which contains multiple questions, but when 2 subsequent questions contain the same answer labels in different positions, the view gets out of sync when selecting one of them and navigating between the 2 questions. 

The radio buttons are contained in a child component which gets the data from the parent component.
The questions are loaded 1 by 1 through a service (faked server call).
The given answers are stored in a array in the Typescript code.

Posted it earlier on SO: https://stackoverflow.com/questions/62910998/angular-radio-buttons-out-of-sync-with-model

## 🔬 Minimal Reproduction
I have a reproduction with and without Angular Material because they both have different behaviour.
Without Material: https://stackblitz.com/edit/angular-9-material-starter-jnbscw
With Material: https://stackblitz.com/edit/angular-9-material-starter-omkvvq

## 🔥 Exception or Error
The view gets out of sync with the model
Without Material: No radio-button is selected.
With Material: The wrong radio-button is selected or multiple (2) radio buttons are selected.

## 🌍  Your Environment

**Angular Version:**
<pre><code>Angular CLI: 9.1.7
Node: 12.18.2
OS: win32 x64

Angular: 9.1.9
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router
Ivy Workspace: Yes

Package                                    Version
--------------------------------------------------------------------
@angular-devkit/architect                  0.901.7
@angular-devkit/build-angular              0.901.7
@angular-devkit/build-optimizer            0.901.7
@angular-devkit/build-webpack              0.901.7
@angular-devkit/core                       9.1.7
@angular-devkit/schematics                 9.1.7
@angular/cdk                               9.2.4
@angular/cli                               9.1.7
@angular/flex-layout                       9.0.0-beta.31
@angular/material                          9.2.4
@angular/material-moment-adapter           9.2.4
@ngtools/webpack                           9.1.7
@nguniversal/module-map-ngfactory-loader   8.2.6
@schematics/angular                        9.1.7
@schematics/update                         0.901.7
rxjs                                       6.5.5
typescript                                 3.8.3
webpack                                    4.42.0</code></pre>

**Comment by gilsdav** [comment]
I don't think this is an Angular issue. You have some wrong logics into your given examples.

1. With material:  `{{ form.controls.question.value | json }}` works. You just never update `unsavedAnswers`.
2. Without material: you have wrong ids. Every elements have id `"i"` because you forgot the binding. `id="i"` => `[id]="i"`. Same for the `for` attribute.

**Comment by Warwizza** [comment]
@gilsdav Thank you for your quick response. 

1. I'm not quite sure what you mean with your first point. At least replacing `{{ unsavedAnswers | json }}` with `{{ form.controls.question.value | json }}` does not resolve the issue.
2. I updated the id and for attributes as you said, but that also does not resolve the issue.

I looked through the logic again and couldn't find something strange that could cause this behaviour. I'm not quite sure that this example should need the use of the `ControlValueAccessor` because to me this seems quite basic functionality that Angular should support. It also works fine when the labels of the radio-button all have different values (or the same value in the same position).

I cannot reproduce it with the `SetTimeout` option for now, but that does feel like a hack and not a robust solution for this problem.

**Comment by AndrewKushnir** [comment]
Hi @Warwizza, thanks for reporting the issue and providing a repro.
It looks like a bug in Forms package potentially related to the `RadioControlValueAccessor` class and `ngFor` logic.
As a possible workaround, you can try adding the `trackBy` into the `ngFor` directive [as shown here](https://angular.io/api/common/NgForOf#description).
Thank you.

**Comment by Warwizza** [comment]
HI @AndrewKushnir, thank you for your response and the workaround.
I tried your workaround and it worked, but we already implemented another workaround where we destroy the child component (with an `ngIf`) and let Angular (re-)initialize it with the new data. This way it's impossible that the `ngFor` has reference to the previous data.
Hopefully this will get fixed soon, so that we can remove our workaround.
